### PR TITLE
test: improve fungible module coverage (system_contract tests)

### PR DESCRIPTION
**Pull request by skosito**
# Description

Adds tests for `system_contract.go` in fungible module to improve coverage.

I think we need separate issue to do some research/discussion on integration vs unit tests for keepers, when and how to write them, since they are a bit mixed up atm and sometimes we need to be able to mock to test how keeper code interacts with some of EvmCall failures. I tried using evm keeper mocks ApplyMessage in some tests, but I think it is very low level call to mock, and sometimes it is awkward since ApplyMessage is used in bunch of places. Also mocking things like abi unpack this way is very unclean (even though those are not so important).
Ideally it would be better to directly mock fungibleKeeper.EvmCall method, maybe we can open additional issue now when coverage is high, to try and write some of these failure tests in that way if possible and avoid mocking too low level calls?

Closes: #1742 

## Type of change

New tests

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [x] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: x/fungible/keeper/system_contract.go [hunk 1]
```diff
@@ -119,11 +123,15 @@ func (k *Keeper) GetUniswapV2FactoryAddress(ctx sdk.Context) (ethcommon.Address,
...
 	}
-	var wzetaResponse AddressResponse
-	if err := sysABI.UnpackIntoInterface(&wzetaResponse, "uniswapv2FactoryAddress", res.Ret); err != nil {
+	var uniswapFactoryResponse AddressResponse
+	if err := sysABI.UnpackIntoInterface(&uniswapFactoryResponse, "uniswapv2FactoryAddress", res.Ret); err != nil {
 		return ethcommon.Address{}, cosmoserrors.Wrapf(types.ErrABIUnpack, "failed to unpack uniswapv2FactoryAddress: %s", err.Error())
 	}
-	return wzetaResponse.Value, nil
+
+	if uniswapFactoryResponse.Value == ethcommon.HexToAddress("0x0") {
```
Comment (lumtis) [inline x/fungible/keeper/system_contract.go:131]: I think we can use `ethcommon.IsHexAddress`, same for below

File: x/fungible/keeper/system_contract_test.go [hunk 2]
```diff
@@ -52,307 +53,760 @@ func TestKeeper_GetWZetaContractAddress(t *testing.T) {
...
+	require.Equal(t, big.NewInt(42), balance)
+}
+
+func TestKeeper_QueryWZetaBalanceOfFails(t *testing.T) {
+	k, ctx, _, _ := keepertest.FungibleKeeper(t)
+	k.GetAuthKeeper().GetModuleAccount(ctx, types.ModuleName)
+
+	// fail if no system contract
+	_, err := k.QueryWZetaBalanceOf(ctx, sample.EthAddress())
+	require.Error(t, err)
```
Comment (lumtis) [inline x/fungible/keeper/system_contract_test.go:118]: I believe we can also check `require.ErrorIs(..., types.ErrContractNotFound)`

**Review by lumtis — COMMENTED** [review]
This looks good. I will have a second look once we merge codecov

The general suggestion would be to change the name of the test function by using only the method name to be tested and give a description of the actual test, like "fail if not deploy" in a sub test description using `t.Run`

File: x/fungible/keeper/system_contract.go [hunk 3]
```diff
@@ -87,6 +87,10 @@ func (k *Keeper) GetWZetaContractAddress(ctx sdk.Context) (ethcommon.Address, er
 	if err := sysABI.UnpackIntoInterface(&wzetaResponse, "wZetaContractAddress", res.Ret); err != nil {
 		return ethcommon.Address{}, cosmoserrors.Wrapf(types.ErrABIUnpack, "failed to unpack wZetaContractAddress: %s", err.Error())
 	}
+
+	if wzetaResponse.Value == ethcommon.HexToAddress("0x0") {
```
Comment (ws4charlie) [inline x/fungible/keeper/system_contract.go:91]: There are many places using `if some_address == ethcommon.Address{}`, should we use a unified pattern?

File: x/fungible/keeper/system_contract.go (hunk 1 above)
Comment (ws4charlie) [inline x/fungible/keeper/system_contract.go:131]: should we unify this check by comparing with a `ethcommon.Address{}`? I don't know which one is better but we do have many places that use `ethcommon.Address{}`

File: x/fungible/keeper/system_contract.go [hunk 4]
```diff
@@ -159,6 +167,10 @@ func (k *Keeper) GetUniswapV2Router02Address(ctx sdk.Context) (ethcommon.Address
 	if err := sysABI.UnpackIntoInterface(&routerResponse, "uniswapv2Router02Address", res.Ret); err != nil {
 		return ethcommon.Address{}, cosmoserrors.Wrapf(types.ErrABIUnpack, "failed to unpack uniswapv2Router02Address: %s", err.Error())
 	}
+
+	if routerResponse.Value == ethcommon.HexToAddress("0x0") {
+		return ethcommon.Address{}, sdkerrors.Wrapf(types.ErrContractNotFound, "uniswap router contract invalid address")
```
Comment (ws4charlie) [inline x/fungible/keeper/system_contract.go:172]: same for check and error wrapping

File: x/fungible/keeper/system_contract.go [hunk 5]
```diff
@@ -265,6 +277,9 @@ func (k *Keeper) QuerySystemContractGasCoinZRC20(ctx sdk.Context, chainid *big.I
 	if err := sysABI.UnpackIntoInterface(&zrc20Res, "gasCoinZRC20ByChainId", res.Ret); err != nil {
 		return ethcommon.Address{}, cosmoserrors.Wrapf(types.ErrABIUnpack, "failed to unpack gasCoinZRC20ByChainId: %s", err.Error())
 	}
+	if zrc20Res.Value == ethcommon.HexToAddress("0x0") {
```
Comment (ws4charlie) [inline x/fungible/keeper/system_contract.go:280]: same?

**Review by brewmaster012 — APPROVED** [review]
LGTM

**Review by lumtis — APPROVED** [review]
LGTM

**Pull request merged**